Ingester crashed with panic: interface conversion: interface {} is nil, not time.Time

**Describe the bug**

One of the ingesters crashed after running for a few days with the following stack trace.

```
level=info ts=2021-01-19T21:01:08.54066371Z caller=shipper.go:334 org_id=system msg="upload new block" id=01EWE756CY88FHFBECGJ0N1EXV " stream="stderr" time="2021-01-19T21:01:08.54100429Z"
level=warn ts=2021-01-19T21:01:08.545704974Z caller=s3.go:417 msg="could not guess file size for multipart upload; upload might be not optimized" name=system/debug/metas/01EWE756CY88FHFBECGJ0N1EXV.json err="unsupported type of io.Reader: *bytes.Reader" " stream="stderr" time="2021-01-19T21:01:08.54594754Z"
panic: interface conversion: interface {} is nil, not time.Time " stream="stderr" time="2021-01-19T21:01:17.873853517Z"
 " stream="stderr" time="2021-01-19T21:01:17.873923909Z"
goroutine 66869545 [running]: " stream="stderr" time="2021-01-19T21:01:17.873932149Z"
github.com/armon/go-metrics/prometheus.(*PrometheusSink).Collect.func1(0x2053280, 0xc04e7f7b00, 0x23c4920, 0xc16be6b100, 0x1) " stream="stderr" time="2021-01-19T21:01:17.873936379Z"
 /__w/cortex/cortex/vendor/github.com/armon/go-metrics/prometheus/prometheus.go:76 +0x1d0 " stream="stderr" time="2021-01-19T21:01:17.873940279Z"
sync.(*Map).Range(0xc0000268f0, 0xc002bc7e90) " stream="stderr" time="2021-01-19T21:01:17.875215843Z"
 /usr/local/go/src/sync/map.go:333 +0x1c5 " stream="stderr" time="2021-01-19T21:01:17.875265125Z"
github.com/armon/go-metrics/prometheus.(*PrometheusSink).Collect(0xc0000268f0, 0xc16b552840) " stream="stderr" time="2021-01-19T21:01:17.875270515Z"
 /__w/cortex/cortex/vendor/github.com/armon/go-metrics/prometheus/prometheus.go:74 +0xd5 " stream="stderr" time="2021-01-19T21:01:17.875274665Z"
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1() " stream="stderr" time="2021-01-19T21:01:17.875278905Z"
 /__w/cortex/cortex/vendor/github.com/prometheus/client_golang/prometheus/registry.go:444 +0x19d " stream="stderr" time="2021-01-19T21:01:17.875282955Z"
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather " stream="stderr" time="2021-01-19T21:01:17.875538402Z"
 /__w/cortex/cortex/vendor/github.com/prometheus/client_golang/prometheus/registry.go:536 +0xe36 " stream="stderr" time="2021-01-19T21:01:17.875792669Z"
level=info ts=2021-01-19T21:01:20.743597534Z caller=main.go:187 msg="Starting Cortex" version="(version=1.6.0, branch=HEAD, revision=2f7d37d)" " stream="stderr" time="2021-01-19T21:01:20.743890232Z"
```

The ingester is crash looping and unable to recover after this failure.

```
level=info ts=2021-01-20T12:53:37.289615838Z caller=head.go:717 org_id=default msg="WAL segment loaded" segment=1046 maxSegment=1327
level=info ts=2021-01-20T12:53:43.036512246Z caller=head.go:717 org_id=default msg="WAL segment loaded" segment=1047 maxSegment=1327
level=info ts=2021-01-20T12:53:46.748127216Z caller=head.go:717 org_id=default msg="WAL segment loaded" segment=1048 maxSegment=1327
level=info ts=2021-01-20T12:53:47.578298028Z caller=signals.go:55 msg="=== received SIGINT/SIGTERM ===\n*** exiting"
level=error ts=2021-01-20T12:53:47.578494764Z caller=cortex.go:391 msg="module failed" module=ingester-service err="context canceled"
level=info ts=2021-01-20T12:53:47.578650788Z caller=memberlist_client.go:527 msg="leaving memberlist cluster"
level=info ts=2021-01-20T12:53:48.08291409Z caller=module_service.go:90 msg="module stopped" module=memberlist-kv
level=info ts=2021-01-20T12:53:48.083179307Z caller=server_service.go:50 msg="server stopped"
level=info ts=2021-01-20T12:53:48.083215029Z caller=module_service.go:90 msg="module stopped" module=server
level=info ts=2021-01-20T12:53:48.083240249Z caller=cortex.go:380 msg="Cortex stopped"
level=error ts=2021-01-20T12:53:48.083333321Z caller=log.go:150 msg="error running cortex" err="failed services\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/__w/cortex/cortex/pkg/cortex/cortex.go:426\nmain.main\n\t/__w/cortex/cortex/cmd/cortex/main.go:189\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:203\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1373"

```

**To Reproduce**
Steps to reproduce the behavior:
1. Start Cortex (version=1.6.0, branch=HEAD, revision=2f7d37d)

**Environment:**
 - Infrastructure: Kubernetes, AWS EKS, S3 storage

**Storage Engine**
- [x] Blocks
- [ ] Chunks

**Additional Context**
<!--  Additional relevant info which can help us debug this issue easily like Logs, Configuration etc. -->

A similar issue was reported in hashicorp/vault#9561 and fixed in armon/go-metrics#109.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -81,6 +81,7 @@
 * [BUGFIX] Alertmanager: don't serve HTTP requests until Alertmanager has fully started. Serving HTTP requests earlier may result in loss of configuration for the user. #3679
 * [BUGFIX] Do not log "failed to load config" if runtime config file is empty. #3706
 * [BUGFIX] Do not allow to use a runtime config file containing multiple YAML documents. #3706
+* [BUGFIX] Memberlist: fixed panic caused by race condition in `armon/go-metrics` used by memberlist client. #3724
 
 ## 1.6.0
 

--- a/go.mod
+++ b/go.mod
@@ -11,7 +11,7 @@ require (
 	github.com/NYTimes/gziphandler v1.1.1
 	github.com/alecthomas/units v0.0.0-20190924025748-f65c72e2690d
 	github.com/alicebob/miniredis v2.5.0+incompatible
-	github.com/armon/go-metrics v0.3.3
+	github.com/armon/go-metrics v0.3.6
 	github.com/aws/aws-sdk-go v1.35.31
 	github.com/bradfitz/gomemcache v0.0.0-20190913173617-a41fca850d0b
 	github.com/cespare/xxhash v1.1.0

